yjit/src/codegen.rs

```diff
@@ -4020,6 +4020,117 @@ fn jit_rb_false(
     true
 }
 
+
+/// Codegen for Kernel#is_a?
+fn jit_rb_kernel_is_a(
+    jit: &mut JITState,
+    ctx: &mut Context,
+    asm: &mut Assembler,
+    ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<IseqPtr>,
+    argc: i32,
+    known_recv_class: *const VALUE,
+) -> bool {
+    if argc != 1 {
+        return false;
+    }
+
+    // If this is a super call we might not know the class
+    if known_recv_class.is_null() {
+        return false;
+    }
+
+    // Important note: This works because we depend on known_recv_class.
+    // The left hand side here will always have the same type because there
+    // is a guard before this method call that ensures that is the case.
+    // We just guard that the right hand side does not change.
+
+    let sample_rhs = jit_peek_at_stack(jit, ctx, 0);
+    let sample_lhs = jit_peek_at_stack(jit, ctx, 1);
+
+    // We are not allowing module here because the module hierachy can change at runtime.
+    if !unsafe { RB_TYPE_P(sample_rhs, RUBY_T_CLASS) } {
+        return false;
+    }
+    let sample_is_a = unsafe { rb_obj_is_kind_of(sample_lhs, sample_rhs) == Qtrue };
+
+    let side_exit = get_side_exit(jit, ocb, ctx);
+    asm.comment("Kernel#is_a?");
+    asm.cmp(ctx.stack_opnd(0), sample_rhs.into());
+    asm.jne(counted_exit!(ocb, side_exit, send_is_a_class_mismatch));
+
+    ctx.stack_pop(2);
+
+    if sample_is_a {
+        let stack_ret = ctx.stack_push(Type::True);
+        asm.mov(stack_ret, Qtrue.into());
+    } else {
+        let stack_ret = ctx.stack_push(Type::False);
+        asm.mov(stack_ret, Qfalse.into());
+    }
+    return true;
+}
+
+/// Codegen for Kernel#instance_of?
+fn jit_rb_kernel_instance_of(
+    jit: &mut JITState,
+    ctx: &mut Context,
+    asm: &mut Assembler,
+    ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<IseqPtr>,
+    argc: i32,
+    known_recv_class: *const VALUE,
+) -> bool {
+    if argc != 1 {
+        return false;
+    }
+
+    // If this is a super call we might not know the class
+    if known_recv_class.is_null() {
+        return false;
+    }
+
+    // Important note: This works because we depend on known_recv_class.
+    // The left hand side here will always have the same type because there
+    // is a guard before this method call that ensures that is the case.
+    // We just guard that the right hand side does not change.
+
+    let sample_rhs = jit_peek_at_stack(jit, ctx, 0);
+    let sample_lhs = jit_peek_at_stack(jit, ctx, 1);
+
+    // Need to ensure this is a class and not a module/Integer/Array etc.
+    if !unsafe { RB_TYPE_P(sample_rhs, RUBY_T_CLASS) } {
```

Comment:
I think this filters out when `Module === rhs` unnecessarily, I think. If we do `!(RB_TYPE_P(sample_rhs, RUBY_T_CLASS) || RB_TYPE_P(sample_rhs, RUBY_T_MODULE))` we can explain it better as filtering out the cases where the C implementation raises (it does [`class_or_module_required(rhs)`](https://github.com/ruby/ruby/blob/847a0df058a4adb60266213cb8db7bb537c1d09e/object.c#L782-L784)).